File: Question/34_Find_FirstandLast_Position.cpp
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        if(nums.empty()) return {-1, -1};
        int left = binarySearchLeft(nums, target, 0, nums.size());
        if(left >= nums.size() || nums[left] != target)
            return {-1, -1};
        int right = binarySearchRight(nums, target, left, nums.size());
        return {left, right};
    }
    int binarySearchLeft(vector<int>& nums, int& target, int left, int right) {
        if(left >= right) return left;

        int mid = (right - left) / 2 + left;
        int ans = 0;
        if(nums[mid] < target) {
            ans = binarySearchLeft(nums, target, mid + 1, right);
        } else {
            ans = binarySearchLeft(nums, target, left, mid);
        }
        return ans;
    }
    int binarySearchRight(vector<int>& nums, int& target, int left, int right) {
        if(left >= right) return left - 1;

        int mid = (right - left) / 2 + left;
        if(nums[mid] <= target) {
            return binarySearchRight(nums, target, mid + 1, right);
        } else {
            return binarySearchRight(nums, target, left, mid);
        }

    }
};

File: Question_Java/34_Find_FirstandLast_Position.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        if (nums.length == 0) return new int[]{-1, -1};

        int begin = search(nums, target, false);
        if (begin >= nums.length || nums[begin] != target) return new int[]{-1, -1};
        int end = search(nums, target, true);
        return new int[]{begin, end - 1};
    }
    private int search(int[] nums, int target, boolean begin) {
        int left = 0;
        int right = nums.length - 1;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if ( (begin && nums[mid] > target) || (!begin && nums[mid] >= target)) right = mid - 1;
            else left = mid + 1;
        }
        return left;
    }
}